_posts/2018-12-18-blog-a-more-complete-front-end-testing-strategy.md

```diff
@@ -0,0 +1,243 @@
+
+
+- UT
+
+Unit testing is a software testing method by which individual units of source code, sets of one or more computer program modules together with associated control data, usage procedures, and operating procedures, are tested to determine whether they are fit for use.
+
+
+
+## What's wrong with the general front-end testing?
+
+When we are evaluating the integrity of a test strategy, we try to validate it with some of the following criteria:
+
+* Pass acceptance criteria
+* Catch bugs as early as possible
+* Faster running speed, especially at the top level of testing
+* Find bugs efficiently
+* Test writing costs and maintenance costs
+* Continuous refactoring risks
+
+A better test strategy should not be as consistent as possible with the above items.
+
+In the general testing strategy mentioned above, the problems posed by the problem, we may not be able to catch bugs as early as possible. Because the E2E test is most likely to run once in a relatively long period of time, rather than every PR or even every code commit, because the e2e run is usually slow and unstable, it is the most expensive automated test to run.
+
+Sometimes we have some integration tests that start a bloated integrated system, maybe it already includes a lot of mock, and you can keep running tests over and over again, but when there are more and more cases of integration testing, we can't even guarantee faster completion within each PR, and in a resource-constrained CI environment, it could be half an hour, or even longer.
+
+As a system becomes more complex, we need a complete testing strategy to tell us what cases failed in these tests, and we were able to catch bugs more efficiently through test reports. Whether it is network instability, back-end server APIs exceptions, front-end domain modules exceptions or UI components exceptions, and so on, we can quickly catch this bugs. Obviously, the general testing strategy can provide a limited amount of lookup bugs help. For example, UT has succeeded, IT has failed, and E2E has failed too. It is difficult for us to analyze clearer information from such test reports.
+
+The cost of writing test code should be balanced with a continuable delivery development model. When the AC definition is clear, it's just that our test code should be able to overwrite the information described by AC. In theory, if all AC is guaranteed by E2E, this will also enable the verification of AC. But obviously, this brings with it a highly unbalanced test instability and inefficient operation.
+
+If the unit tests are adequate enough, will this ensure that our AC can be accepted and become viable? This should depend on the maintenance cost of the unit test, our each refactoring will modify the code, as well as the unit tests involved in the code, then the code variability of the unit test means that we should need a higher level of testing than UT to ensure that we refactor the code logic.
```

Comment:
'our each refactoring...' again is not good English, change it to 'each time we refactor our code, we will have to modify the unit tests involved in these code as well, which means we need a higher level of test to ensure the quality and correctness of these code, despite their fast changing nature.'